src/runtime/database.cpp

```diff
@@ -1407,49 +1319,142 @@ static std::vector<FileDependency> get_all_file_dependencies(const Database *db,
-  return find_all(this, imp->find_job);
+  return out;
 }
 
-std::vector<JobReflection> Database::files_matching(const std::string &glob, int use) {
-  const char *why = "Could not bind args";
-  bind_string(why, imp->find_owner, 1, glob);
-  bind_integer(why, imp->find_owner, 2, use);
-  return find_all(this, imp->find_owner);
-}
+std::string collapse_and(const std::vector<std::vector<std::string>> &ands) {
+  if (ands.empty()) {
+    return "";
+  }
 
-std::vector<JobReflection> Database::tags_matching(const std::string &uri_glob,
-                                                   const std::string &content_glob) {
-  const char *why = "Could not bind args";
-  bind_string(why, imp->find_tag, 1, uri_glob);
-  bind_string(why, imp->find_tag, 2, content_glob);
-  return find_all(this, imp->find_tag);
+  if (ands.size() == 1) {
+    return collapse_or(ands.front());
+  }
+
+  std::string out = "";
+
+  for (std::vector<std::vector<std::string>>::size_type i = 0; i < ands.size() - 1; i++) {
+    out += collapse_or(ands[i]) + "\nAND\n\t";
+  }
+
+  out += collapse_or(ands.back());
+
+  return out;
 }
 
-std::vector<JobReflection> Database::tags_matching(const std::string &glob) {
-  std::vector<std::string> parts = wcl::split_by_fn(
-      '=', glob.begin(), glob.end(), [](auto a, auto b) { return std::string(a, b); });
-  if (parts.size() != 2) {
-    std::cerr << "Tag query must have exactly one '=' per tag" << std::endl;
+std::vector<JobReflection> Database::matching(
+    const std::vector<std::vector<std::string>> &and_or_filters) {
+  // This query creates a subtable of the following shape:
+  //
+  // | job_id | label | run_id | use_id | endtime | commandline | status |       input_files   |
+  // output_files  |       tags       |
+  // --------------------------------------------------------------------------------------------------------------------------------
+  // |    1   |  foo  |   1    |    1   |  1234   | ls lah .    |   0    | <d>a.txt<d>b.txt<d> |
+  // <d>c.o<d>d.o<d> | <d>a=b<d>c=d<d>  | |    1   |  foo  |   1    |    1   |  1234   | ls lah . |
+  // 0    | <d>a.txt<d>b.txt<d> | <d>c.o<d>d.o<d> | <d>a=b<d>c=d<d>  |
+  //
+  // The subtable is constructed by joining the jobs table with the minimal set of other dependent
+  // tables with the following extra processing.
+  // 1. files table is split into 2 columns, input_files & output_files
+  // 2. tags are flattened from two columns (uri, content) to one column (tags) with a = separator
+  // 3. input_files, output_files, and tags are group_concat'd into a single row per job. <d> is
+  //    used as a deliminator between each value. The deliminator is also places at the beginning
+  //    and end of each row so that queries don't need to special case the first/last entry.
+  //
+  // Any inspection flag/user code may add any WHERE expression conditions to the main query using
+  // the columns of the subtable for fine grain filters.
```

Comment:
How did you select the columns to include? Is it just the minimum set needed to implement the previous high-level database inspection filters?

Reply:
yep